compactor/src/compact.rs

```diff
@@ -356,28 +366,132 @@ impl Compactor {
+                            .entry(pf.id)
+                            .or_insert_with(|| ParquetFileWithTombstone {
+                                data: Arc::new(pf),
+                                tombstones: vec![],
+                            });
+                        file.add_tombstones(vec![ts.clone()]);
+                    }
+                }
+
+                // Group parquet_file_with_tombstones into partition and add it into the partitions
+                let mut table_partittions: BTreeMap<
+                    PartitionId,
+                    Vec<Arc<ParquetFileWithTombstone>>,
+                > = BTreeMap::new();
+                for (_, pf) in parquet_file_with_tombstones {
+                    let list = table_partittions
+                        .entry(pf.data.partition_id)
+                        .or_insert_with(Vec::new);
+                    list.push(Arc::new(pf));
+                }
+
+                let ps = table_partittions
+                    .into_iter()
+                    .map(
+                        |(partition_id, files_with_tombstones)| PartitionWithParquetFiles {
+                            sequencer_id: *sequencer_id,
+                            table_id,
+                            partition_id,
+                            files_with_tombstones,
+                        },
+                    )
+                    .collect::<Vec<_>>();
+
+                partitions.extend(ps);
+            }
+        }
+
+        Ok(partitions)
+    }
+
     /// Runs compaction in a partition resolving any tombstones and compacting data so that parquet
     /// files will be non-overlapping in time.
+    /// If no files are provided, the files will be gotten from the catalog
     pub async fn compact_partition(
         &self,
         partition_id: PartitionId,
         compaction_max_size_bytes: i64,
+        files: Option<Vec<Arc<dyn FileMeta>>>,
     ) -> Result<()> {
         info!("compacting partition {}", partition_id);
         let start_time = self.time_provider.now();
 
-        let parquet_files = self
-            .catalog
-            .repositories()
-            .await
-            .parquet_files()
-            .list_by_partition_not_to_delete(partition_id)
-            .await
-            .context(ListParquetFilesSnafu)?;
+        let parquet_files = match files {
+            // files provided, compact them
+            Some(files) => files,
+            None => {
+                // files not provided, get them from the catalog
+                let files = self
+                    .catalog
+                    .repositories()
+                    .await
+                    .parquet_files()
+                    .list_by_partition_not_to_delete(partition_id)
+                    .await
+                    .context(ListParquetFilesSnafu)?;
+
+                let files = files
+                    .iter()
+                    .map(|f| Arc::clone(&(*f)) as Arc<dyn FileMeta>)
```

Comment:
This can be a little simpler, `f` is already `&Arc`:

```suggestion
                    .map(|f| Arc::clone(f) as Arc<dyn FileMeta>)
```

I wish there was a way to coerce this whole `Vec` at once, but I don't know of a way :(